refactor & log statistics from txn coordinator

tbg opened the pull request:
the refactoring is mostly for transparency and to make sure that the transaction record stored in the txnMetadata is updated appropriately, as this is relevant for the stats.

> txn coordinator: 157.60 txn/sec, 92.13/7.87/0.00 %cmmt/abrt/abnd, 16ms/178ms/5.002s avg/σ/max duration, 0.3/0.7/5.0 avg/σ/max restarts (788 samples)

petermattis commented on kv/txn_coord_sender.go:
```diff
@@ -35,8 +35,11 @@ import (
 	"github.com/cockroachdb/cockroach/util/hlc"
 	"github.com/cockroachdb/cockroach/util/log"
 	gogoproto "github.com/gogo/protobuf/proto"
+	"github.com/montanaflynn/stats"
```
We've got this crazy metrics system in `util/metrics`. Is it not appropriate for this usage?

tschottdorf replied:
if we needed an excuse for using it, we could probably do it here, but I remember that we weren't even sure whether it was our best bet and I'm not too inclined to get sidetracked.

petermattis commented on kv/txn_coord_sender.go:
```diff
@@ -507,72 +598,127 @@ func (tc *TxnCoordSender) updateResponseTxn(argsHeader *proto.RequestHeader, rep
+	if _, ok := tc.txns[id]; !ok {
+		return
+	}
+	tc.txnStats.durations = append(tc.txnStats.durations, float64(tc.clock.PhysicalNow()-txnMeta.firstUpdateNanos))
```
I'm not seeing a limit anywhere on the size of `durations` and `restarts`. Am I missing it?

tschottdorf replied:
you mean a limit in terms of slice length? The durations/restarts are reset every time the stats are dumped, if memory churn is a concern we could just have two of each and reuse them in turn.

petermattis replied:
Ah, missed that the `txnStats` object was completely cleared. I was just searching for `durations`.

petermattis commented on kv/txn_coord_sender.go:
```diff
@@ -220,9 +244,62 @@ func NewTxnCoordSender(wrapped client.Sender, clock *hlc.Clock, linearizable boo
+			dDev := time.Duration(stats.StdDevP(curStats.durations))
+			rMax := stats.Max(curStats.restarts)
+			rMean := stats.Mean(curStats.restarts)
+			rDev := stats.StdDevP(curStats.restarts)
```
This feels a bit ad-hoc. I think we want to have more a more generalized notion of stat counter. I thought `util/metrics` would provide that. At Square we used http://godoc.org/github.com/rcrowley/go-metrics (or something like it, can't recall exactly).

tschottdorf replied:
I agree that we'll want something more sophisticated but there's also value in not making use of the next best complicated thing right away. The stats this PR outputs are very useful immediately (even if homegrown) for performance tuning and the rest can be done in a follow-up (I'd also like to see how exactly this is going to play with timeseries).

tschottdorf replied:
http://godoc.org/github.com/rcrowley/go-metrics#CaptureRuntimeMemStats

that's stuff that's relevant for the corresponding (homegrown) code in `server/node` so I could see us wanting to use that framework. But I haven't done any research yet. We should probably figure that out soon.

petermattis replied:
Yeah, I agree that there is immediate value to this stuff. How about adding a TODO indicating that this needs to be generalized?

tschottdorf replied:
done.

petermattis commented:
LGTM

tbg committed:
refactor & log statistics from txn coordinator

the refactoring in txn_coordinator is mostly for
transparency and to make sure that the transaction
record stored in the txnMetadata is updated appropriately,
as this is relevant for the stats.

txn coordinator: 157.60 txn/sec, 92.13/7.87/0.00 %cmmt/abrt/abnd, 16ms/178ms/5.002s avg/σ/max duration, 0.3/0.7/5.0 avg/σ/max restarts (788 samples)

spencerkimball commented:
Is there a TODO in there for @mrtracy to implement this as a time series recorder?

tschottdorf commented:
yes, pretty sure I added one.

tschottdorf commented on kv/txn_coord_sender.go:
```diff
@@ -220,9 +244,64 @@ func NewTxnCoordSender(wrapped client.Sender, clock *hlc.Clock, linearizable boo
+// success rates, durations, ...).
+// TODO(tschottdorf): Use a proper metrics subsystem for this (+the store-level
+// stats).
+// TODO(mrtracy): Add this to TimeSeries.
```
@spencerkimball right here.

spencerkimball replied:
:+1:
